payments/models.py

```diff
@@ -212,15 +215,23 @@ def release(self):
     def refund(self, amount=None):
         if self.status != PaymentStatus.CONFIRMED:
             raise ValueError("Only charged payments can be refunded.")
-        if amount:
-            if amount > self.captured_amount:
-                raise ValueError(
-                    "Refund amount can not be greater then captured amount"
-                )
-            provider = provider_factory(self.variant, self)
-            amount = provider.refund(self, amount)
-            self.captured_amount -= amount
-        if self.captured_amount == 0 and self.status != PaymentStatus.REFUNDED:
+        if amount and amount > self.captured_amount:
+            raise ValueError("Refund amount can not be greater then captured amount")
+        provider = provider_factory(self.variant, self)
+        amount = provider.refund(self, amount)
+        # If the initial amount is None, the code above has no chance to check whether
+        # the actual amount is greater than the captured amount before actually
+        # performing the refund. But since the refund has been performed already,
+        # raising an exception would just cause inconsistencies. Thus, logging an error.
```

Comment:
Under what conditions would the refund amount be larger than the captured amount?

Reply:
@WhyNotHugo, since 2021, every (our) PayPal payment model has the captured amount equal to zero due to https://github.com/jazzband/django-payments/issues/309. Thus, every attempt to refund any of those payments (no matter if `amount` is `None` or not) has the refund amount greater than the captured amount. (And such asynchrony between reality and models can easily happen due to other bugs as well.)

Reply:
In that case, won't you be logging an error for every single refund below?

Reply:
Yes, indeed. There will be fewer and fewer such log messages thanks to your recent merge of https://github.com/jazzband/django-payments/pull/412 though.
However, this is just one case that can eventually be avoided by fixing the database since I know about the wrong database data. There are possibly other users of django-payments that are not aware of this. And there are possible other cases when the captured amounts do not match the reality.